### ISSUE TITLE
how to set 𝜇 in grpo paper?

### ISSUE THREAD

**Issue by Chandler-Bing** [volcengine/verl#1008]
[dataloader = batch.split(self.config.ppo_mini_batch_size)](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L254)  

Why do we split a batch into mini-batches here? Moreover, each mini-batch updates the model. The `old_log_probs` always corresponds to the actor model in the initial state of the mini-batch. How is this different from the implementation of  𝜇  in the grpo?

**Comment by 0x404** [comment]
I think the `ppo_mini_batch_size` is used for GPU memory management since verl can't do all update at a time with the limited GPU memory. and [`ppo_epochs`](https://github.com/volcengine/verl/blob/c9e3c57cf88378ca39817881b65d0531e2a7ffdf/verl/trainer/config/ppo_trainer.yaml#L46) is corresponding to the u parameter in the GRPO algo.

**Comment by Chandler-Bing** [comment]
> I think the `ppo_mini_batch_size` is used for GPU memory management since verl can't do all update at a time with the limited GPU memory. and [`ppo_epochs`](https://github.com/volcengine/verl/blob/c9e3c57cf88378ca39817881b65d0531e2a7ffdf/verl/trainer/config/ppo_trainer.yaml#L46) is corresponding to the u parameter in the GRPO algo.

according to [grad_norm = self._optimizer_step()](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L343) actor model update every  `ppo_mini_batch_size` , it's more like a `𝜇` to me. if `ppo_epochs` corresponding to the `𝜇` ,when `ppo_epochs=1` , `old_log_probs` supposed to be same as `log_probs`, and so clip never happens, but the code above seems a different version?

**Comment by 0x404** [comment]
> if ppo_epochs corresponding to the 𝜇 ,when ppo_epochs=1 , old_log_probs supposed to be same as log_probs

let's say `ppo_epochs=1`, `batch_size=32` and `ppo_mini_batch_size=8`. in this setting, we will have to do 4 ppo_mini iteration. in each iteration, we will first calculate the `log_probs` from the actor and then calculate the loss, the we backward the loss and update the parameter of the actor. yes, in the first iteration, `log_probs` is the same as `old_log_probs`, but since actor will be updated in the first iteration, after that, `log_probs` may no the same as `old_log_probs`.

<img width="771" alt="Image" src="https://github.com/user-attachments/assets/c72012e2-4a3c-4e3c-9d56-ea6893e12a05" />

In the GRPO, each iteration of u should use all data samples to update the actor, and of course if u = 1, the clip will no happen. in this case, I think when we set `ppo_mini_batch_size = train_batch_size`, the `ppo_epochs` is correspoding to u.

this is my personal understanding, what do you think?

**Comment by Chandler-Bing** [comment]
> > if ppo_epochs corresponding to the 𝜇 ,when ppo_epochs=1 , old_log_probs supposed to be same as log_probs
> 
> let's say `ppo_epochs=1`, `batch_size=32` and `ppo_mini_batch_size=8`. in this setting, we will have to do 4 ppo_mini iteration. in each iteration, we will first calculate the `log_probs` from the actor and then calculate the loss, the we backward the loss and update the parameter of the actor. yes, in the first iteration, `log_probs` is the same as `old_log_probs`, but since actor will be updated in the first iteration, after that, `log_probs` may no the same as `old_log_probs`.
> 
> <img alt="Image" width="771" src="https://private-user-images.githubusercontent.com/18774274/432606599-c72012e2-4a3c-4e3c-9d56-ea6893e12a05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2ODE4MjgsIm5iZiI6MTc0NDY4MTUyOCwicGF0aCI6Ii8xODc3NDI3NC80MzI2MDY1OTktYzcyMDEyZTItNGEzYy00ZTNjLTlkNTYtZWE2ODkzZTEyYTA1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE1VDAxNDUyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwN2E2NzExYWYxOWNkZDYyNWI2ZmFjYTI4ZjUyMGYyNzhlYmMzYmNkZDY5NzNjMDQwZWU1ODQzMzllYzM0ZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RQZ8Tp7JqFd2zeNMLbbQntrPIt1lHX_TRQfbCVyVs4k">
> In the GRPO, each iteration of u should use all data samples to update the actor, and of course if u = 1, the clip will no happen. in this case, I think when we set `ppo_mini_batch_size = train_batch_size`, the `ppo_epochs` is correspoding to u.
> 
> this is my personal understanding, what do you think?

agreed, there are slight differences between verl and original grpo paper, maybe it's a better version

**Comment by fengxiang777** [comment]
> I think the `ppo_mini_batch_size` is used for GPU memory management since verl can't do all update at a time with the limited GPU memory. and [`ppo_epochs`](https://github.com/volcengine/verl/blob/c9e3c57cf88378ca39817881b65d0531e2a7ffdf/verl/trainer/config/ppo_trainer.yaml#L46) is corresponding to the u parameter in the GRPO algo.

Thank you for your patience in responding. However, I am curious about your statement that ``ppo_mini_batch`` is for memory management. If gradient accumulation is used, is the memory consumption still affected by the ``train_batch_size``?

**Comment by Kirrito-k423** [comment]
定时清理检查结果：这个 issue 最近一次由当前运行账号以外账号产生的公开回复或公开引用是 `2025-06-05T01:34:44Z`。本轮检查日期为 2026-07-04，该时间已经超过 6 个月。检查来源包括 issue comments API `https://api.github.com/repos/verl-project/verl/issues/1008/comments?per_page=100` 和 timeline API `https://api.github.com/repos/verl-project/verl/issues/1008/timeline?per_page=100`。

按照本仓库的定时 issue 清理标准，我会先留下这条说明并关闭该 issue。若作者或社区成员仍能提供新的复现、证据或需求背景，请继续评论请求重开，或创建新的 issue 并引用这里的上下文。

---

Scheduled cleanup check: the latest public reply or public reference from an account other than the current automation account was `2025-06-05T01:34:44Z`. The check date for this run is 2026-07-04, so that activity is more than 6 months old. The checked public sources were the issue comments API `https://api.github.com/repos/verl-project/verl/issues/1008/comments?per_page=100` and the timeline API `https://api.github.com/repos/verl-project/verl/issues/1008/timeline?per_page=100`.

Following the scheduled issue cleanup policy, I am leaving this note first and then closing the issue. If the author or community can still provide a fresh reproduction, evidence, or requirement context, please comment to request reopening or open a new issue that links back to this one.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
